### PR TITLE
Healthcheck managed for all related services

### LINKED ISSUE #5231: Failing health check but still available
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [X] No

### What did you do?
I have a small nginx that I ocassionally take "offline" by deliberately failing the health check, which works with traefik 1.7.12 but no longer in v2.0.

I have configured the health check as below and *see* the health check in the nginx log giving a `404`, but even then, the server is still available. Am I missing something?

### What did you expect to see?

Failing health check takes service offline (works in 1.7.12)

### What did you see instead?

Service still available, `serverStatus` in `/api/rawdata` still `UP`

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.0-beta1
Codename:     faisselle
Go version:   go1.12.7
Built:        2019-07-19T16:04:34Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?
Docker provider is used, labels:
```yaml
    labels:
      - "traefik.docker.network=web"
      - "traefik.enable=true"
      - "traefik.http.routers.stuffhttps.rule=Host(`domain`) || Host(`domain2`)"
      - "traefik.http.routers.stuffhttps.entrypoints=https"
      - "traefik.http.routers.stuffhttps.tls.certresolver=default"
      - "traefik.http.routers.stuffhttps.middlewares=sec@file"

      - "traefik.http.routers.stuffhttp.middlewares=redirect"
      - "traefik.http.routers.stuffhttp.rule=Host(`domain`) || Host(`domain2`)"
      - "traefik.http.routers.stuffhttp.entrypoints=http"
           
      - "traefik.http.middlewares.redirect.headers.sslredirect=true"
      
      - "traefik.http.services.stuff.loadbalancer.healthcheck.path=/health/health"
      - "traefik.http.services.stuff.loadbalancer.healthcheck.interval=15s"
      - "traefik.http.services.stuff.loadbalancer.healthcheck.timeout=1s"
```

nginx log example:

```
172.29.0.3 - - [20/Aug/2019:06:08:14 +0000] "GET /health/health HTTP/1.1" 404 153 "-" "Go-http-client/1.1" "-"                                                                                          
2019/08/20 06:08:29 [error] 6#6: *89 open() "/usr/share/nginx/html/health/health" failed (2: No such file or directory), client: 172.29.0.3, server: localhost, request: "GET /health/health HTTP/1.1", 
host: "172.29.0.6:80"
```

### PR DESCRIPTION
**Pull request by jbdoumenjou**
### What does this PR do?

Iterates on all health check load balancers to apply the status from servers.

### Motivation

Fixes #5231 

### More

- [x] Added/updated tests
- [ ] ~~Added/updated documentation~~

Co-authored-by: Mathieu Lonjaret <mathieu.lonjaret@gmail.com>

### CODE REVIEW

File: pkg/healthcheck/healthcheck.go [hunk 1]
```diff
@@ -281,3 +281,41 @@ func (lb *LbStatusUpdater) UpsertServer(u *url.URL, options ...roundrobin.Server
 	}
 	return err
 }
+
+// BalancerHandlers is a list of BalancerHandler(s) that implements BalancerHandler.
+type BalancerHandlers []BalancerHandler
+
+func (b BalancerHandlers) ServeHTTP(w http.ResponseWriter, req *http.Request) {
```
Comment (mpl) [inline pkg/healthcheck/healthcheck.go]: This is weird, but I don't know yet if it is a problem. I need to think about it.

Comment (mpl) [reply pkg/healthcheck/healthcheck.go]: alternative implementation that removes the need for that empty ServeHTTP: 
https://github.com/mpl/traefik/tree/healthcheck-lb
(clean commit on top of the last one in this PR)

Comment (mpl) [reply pkg/healthcheck/healthcheck.go]: I pushed that commit to the PR, as agreed by chat with @jbdoumenjou

File: pkg/server/service/service.go [hunk 2]
```diff
@@ -242,7 +243,7 @@ func (m *Manager) LaunchHealthCheck() {
 	healthcheck.GetHealthCheck().SetBackendsConfiguration(context.Background(), backendConfigs)
 }
 
-func buildHealthCheckOptions(ctx context.Context, lb healthcheck.BalancerHandler, backend string, hc *dynamic.HealthCheck) *healthcheck.Options {
+func buildHealthCheckOptions(ctx context.Context, lb healthcheck.Balancers, backend string, hc *dynamic.HealthCheck) *healthcheck.Options {
```
Comment (juliens) [inline pkg/server/service/service.go]: You should use the `Balancer` interface here.

File: pkg/server/service/service.go [hunk 3]
```diff
@@ -53,10 +53,14 @@ type Manager struct {
...
 	defaultRoundTripper http.RoundTripper
-	balancers           map[string][]healthcheck.BalancerHandler
-	configs             map[string]*runtime.ServiceInfo
-	api                 http.Handler
-	rest                http.Handler
+	// balancers is the map of all Balancers, keyed by service name.
+	// There is one Balancer per service handler, and there is one service handler per reference to a service
+	// (e.g. if 2 routers refer to the same service name, 2 service handlers are created),
+	// which is why there is not just one Balancer per service name.
+	balancers map[string][]healthcheck.Balancer
```
Comment (juliens) [inline pkg/server/service/service.go]: You should use the `healthcheck.Balancers` instead of `[]healthcheck.Balancer`.

**Comment by strarsis** [comment]
Can this be released as docker image to dockerhub so I can test it?

**Comment by ldez** [comment]
The image: `containous/experimental-v2.0`

**Comment by strarsis** [comment]
@ldez: Sorry for the OT 🐱 
So I tried `containous/experimental-v2.0` but when I start two docker containers, 
each with labels for using same frontend with different hostnames, 
both are listed as down as before and I can't reach either of them.
Only when one container is up, it can be reached by its hostname.

**Comment by ldez** [comment]
Could you go to the [community forum](https://community.containo.us/) and provide a docker-compose to reproduce. Thank you.

**Comment by strarsis** [comment]
@ldez: Sure! You can find the topic here: https://community.containo.us/t/all-containers-down-with-more-than-one-container-per-frontend/3218

**Pull request merged**